sashimi-webapp/src/logic/filemanager/file.js

```diff
@@ -76,58 +95,80 @@ File.prototype.load = function load() {
  *
  * @param {Folder} folder default: currentFolder
  * @return {}
- */
+
 File.prototype.copy = function copy(folder) {
-  console.log('file.copy');
-  // Case 1a: file exist, valid folder, file copied
-  // Case 1b: file exist, invalid folder, file not copied (error)
-  // Case 2: file does not exist (not copied)
-  // Case 3: file exist, folder not specified, default: current folder used
+
 };
+*/
 
 /**
  * Move file to a specified folder
  *
- * @param {Folder} folder
- * @return {}
+ * @param {Folder} destFolder
+ * @return {Promise}
+ */
+File.prototype.move = function move(destFolder) {
+  return new Promise((resolve, reject) => {
+    if (isCurrentFolder(destFolder)) {
+      reject(MOVE_SAME_FOLDER_ERROR_MSG);
+    }
+
+    if (isInvalidFolder(destFolder)) {
+      reject(MOVE_INVALID_FODLER_ERROR_MSG);
+    }
+
+    resolve();
+  })
+  .then(() => storage.moveFile(this.id, destFolder.id))
+  .then(() => {
+    destFolder.childFileList.push(this);
+    const index = this.parentFolder.childFileList.indexOf(this);
+    this.parentFolder.childFileList.splice(index, 1);
+    this.parentFolder = destFolder;
+  });
+};
+
+/**
+ * Rename file
+ *
+ * @param {String} newFileName
+ * @return {Promise}
  */
-File.prototype.move = function move() {
-  console.log('file.move');
-  // Changing file path
-  // Case 1a: file exist, valid folder, file moved
-  // Case 1b: file exist, invalid folder, file not moved (error)
-  // Case 1c: file exist, folder is where file currently reside in (not moved)
-  // Case 2: file does not exist (not moved)
+File.prototype.rename = function rename(newFileName) {
+  return new Promise((resolve, reject) => {
+    if (hasSameFileName(newFileName)) {
```

Comment:
Useful read to understand some of the things that happens here
http://stackoverflow.com/questions/80084/in-javascript-why-is-the-this-operator-inconsistent#80478

```javascript
File.prototype.rename = function rename(newFileName) {
  // at this point, the `this` is referred to the File instance, because `rename` is invoked as a method attached to `File` object.

  return new Promise((resolve, reject) => {
    // Here we pass a function into the Promise constructor as a callback.
    // Thus we consider this callback function will be "invoked as a function",
    //   which means that, the `this` inside this function is equal to the context where 
    //   this callback function is called (inside `File.prototype.rename`)
    //   `this` === the File object

    // However `hasSameFileName` is called inside the new Promise context
    //   (e.g. callsite is Promise), which has a `this` === `undefined`. 
    // Therefore, you are not able to access the `File` object using the `this` 
    //   keyword inside `hasSameFileName`.
    if (hasSameFileName(newFileName)) {
      reject(RENAME_ERROR_MSG);
    }

    resolve();
  })
  // This part is still within the same context where the `this` is referred to the `File` object.
  .then(() => storage.renameFile(newFileName, this.id))
  .then(() => {
    const oldFileName = this.name;
    this.name = newFileName;
    this.path = this.path.replace(oldFileName, newFileName);
  });
});

```


I am not exactly fluent with the concept about `this` also. So I may not have explained it correctly.
This article could potentially help you understand more about the behaviour of javascript's `this`
https://github.com/getify/You-Dont-Know-JS/blob/master/this%20%26%20object%20prototypes/ch1.md